Mark addresses as used while discovering them.

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#2032 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- dd3cff59cef4a66bb792adda6491f54b8ea8ca23
  :round_pushpin: **mark addresses as used as they are discovered**
    The main issue of the current list address handler is that it loads in memory and crawls the ENTIRE transaction history of the wallet. This is done merely to know whether addresses have been used or not. This is unnecessary for random addresses because used and unused addresses are already separated in two different Maps. For sequential wallets it's a bit more subtle but, we can mark addresses as "Used" during discovery and store this information in the database.

- 61bbd7b3f0755386a815490868e2e556e13a8e8d
  :round_pushpin: **add two properties for Random & Sequential wallets showing the effect of IsOurs on the address state**
  
- eed460c03a8ce2f5093e59f7d58acf3422653d49
  :round_pushpin: **fix random address import to not mark every imported address as 'Used'**
  

# Comments

<!-- Additional comments or screenshots to attach if any -->

- [ ] **NOTE 1**: This change will require a database migration forcing wallets to re-synchronize from scratch. 

- [x] ~~**NOTE 2**: I haven't yet fixed unit tests which need some adjustment since some of the interfaces have changed.~~

- [ ] **NOTE 3**: I'll try measuring and comparing the behavior with a stress-wallet that owns many addresses.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
mark addresses as used as they are discovered

  The main issue of the current list address handler is that it loads in memory and crawls the ENTIRE transaction history of the wallet. This is done merely to know whether addresses have been used or not. This is unnecessary for random addresses because used and unused addresses are already separated in two different Maps. For sequential wallets it's a bit more subtle but, we can mark addresses as "Used" during discovery and store this information in the database.

KtorZ committed:
add two properties for Random & Sequential wallets showing the effect of IsOurs on the address state

KtorZ committed:
fix random address import to not mark every imported address as 'Used'

rvl commented on lib/core-integration/src/Test/Integration/Framework/Request.hs:
```diff
@@ -176,23 +183,38 @@ request ctx (verb, path) reqHeaders body = do
+    -- separator), and decoding them as a JSON list.
+    decodeOne = Aeson.decodeStrict . mconcat
+    decodeStream chunks = do
+        xs <- traverse (Aeson.decodeStrict @Aeson.Value . B8.init) chunks
```
I don't think you can necessarily assume that the chunks are split nicely at the newline.
The OS might interrupt the socket read early, and you will have a line in two pieces, neither of which are well-formed json.

This streaming change is worthy of its own PR, where the details can be sorted out.

I like the idea of streaming in general, because it removes one case of unbounded memory usage.

But on the other hand, if you have 1 million addresses, that would be perhaps 80MB of encoded json, as an indication of size. So the memory cost of not streaming is smallish. Without streaming there is a delay while the server is reading the list from sqlite, but unless the clients also have stream processing, so what?

KtorZ replied:
@rvl actually.. there's one test failing because of this which shows that sometimes, there are chunks that only contains half of an address (it fails on the "import 15.000 addresses and list them" scenario, I haven't noticed this below 100 addresses). 

So yeah, this require some additional thoughts maybe. I'll split the streaming approach into a separate PR.

rvl reviewed (commented):
I wholly approve of the change to mark sequential addresses as used once they are discovered, so that this information is easily known when listing addresses.

I also like the idea of streaming responses, but the topic needs further discussion in a separate PR.

KtorZ committed:
add migration for new 'status' column in address state

  In the end, migrating an existing database was quite easy given the right SQL incantation. We need to mark any 'known' address as used where a known address is simply an address that appear in one of the wallet known txout. We don't care in the end if it's an outgoing or incoming transaction. Just seeing an address there immediately qualifies it as 'used'.

rvl reviewed (approved):
Migration looks OK

rvl committed:
Sqlite: Column order matters to persistent auto-migrations
